### PR TITLE
108269 extend labling endpoints

### PR DESCRIPTION
**Pull request by janid1967**
Final PR for [AB#108269](https://statoildeveloper.visualstudio.com/b6d97460-2e53-47ae-aad4-b41396d90828/_workitems/edit/108269)
The main task of PR is a new endpoint for handling which label entities a label is available for (UpdateLabelAvailableForCommand and handling of it)

During this work, yet a renaming of the EntityTypeWithLabels enum and use of such emerged as necessary. Hence Label.EntitiesWithLabel is now Label.AvailableFor

### CODE REVIEW

File: src/Equinor.ProCoSys.Completion.Command/LabelCommands/UpdateLabelAvailableFor/UpdateLabelAvailableForCommandValidator.cs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+        RuleFor(command => command)
+            .MustAsync((command, cancellationToken) => BeAnExistingLabelAsync(command.Text, cancellationToken))
+            .WithMessage(command => $"Label does not exist! Guid={command.Text}");
+
+        RuleForEach(command => command.AvailableFor)
+            .MustAsync((_, entityType, _, cancellationToken)
+                => BeAnExistingLabelEntityAsync(entityType, cancellationToken))
+            .WithMessage((_, entityType) => $"Label entity does not exist! Type={entityType}");
+
```
Comment (jsoi-eq) [inline src/Equinor.ProCoSys.Completion.Command/LabelCommands/UpdateLabelAvailableFor/UpdateLabelAvailableForCommandValidator.cs:26]: We are allowing updates to a voided Label?

File: src/Equinor.ProCoSys.Completion.Command/LabelCommands/UpdateLabelAvailableFor/UpdateLabelAvailableForCommandHandler.cs [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        _labelEntityRepository = labelEntityRepository;
+        _unitOfWork = unitOfWork;
+        _logger = logger;
+    }
+
+    public async Task<Result<Unit>> Handle(UpdateLabelAvailableForCommand request, CancellationToken cancellationToken)
+    {
+        var label = await _labelRepository.GetByTextAsync(request.Text, cancellationToken);
+
+        foreach (var entityType in request.AvailableFor)
```
Comment (jsoi-eq) [inline src/Equinor.ProCoSys.Completion.Command/LabelCommands/UpdateLabelAvailableFor/UpdateLabelAvailableForCommandHandler.cs:36]: I struggled a lot to parse this code in my head.
Following is an attempt to split it into smaller pieces with descriptive names.

    public async Task<Result<Unit>> Handle(UpdateLabelAvailableForCommand request, CancellationToken cancellationToken)
    {
        var existingLabel = await _labelRepository.GetByTextAsync(request.Text, cancellationToken);

        await AddToAvailableFor(request, cancellationToken, existingLabel);

        RemoveFromAvailableFor(request, existingLabel);

        await _unitOfWork.SaveChangesAsync(cancellationToken);

        _logger.LogInformation("Usage for Label {Label} updated", request.Text);

        return new SuccessResult<Unit>(Unit.Value);
    }


    private async Task AddToAvailableFor(UpdateLabelAvailableForCommand request, CancellationToken cancellationToken,
        Label existingLabel)
    {
        foreach (var entityTypeFromRequest in request.AvailableFor)
        {
            if (EntityTypeNotInExistingAvailableFor(entityTypeFromRequest, existingLabel.AvailableFor))
            {
                var labelEntity = await _labelEntityRepository.GetByTypeAsync(entityTypeFromRequest, cancellationToken);
                existingLabel.MakeLabelAvailableFor(labelEntity);
            }
        }
    }

    private static void RemoveFromAvailableFor(UpdateLabelAvailableForCommand request, Label existingLabel)
    {
        for (var i = existingLabel.AvailableFor.Count - 1; i >= 0; i--)
        {
            var existingLabelEntity = existingLabel.AvailableFor.ElementAt(i);
            if (!request.AvailableFor.Contains(existingLabelEntity.EntityType))
            {
                existingLabel.RemoveLabelAvailableFor(existingLabelEntity);
            }
        }
    }

    private static bool EntityTypeNotInExistingAvailableFor(EntityTypeWithLabels entityType, ICollection<LabelEntity> availableFor) 
        => availableFor.All(e => e.EntityType != entityType);

File: src/tests/Equinor.ProCoSys.Completion.Command.Tests/LabelCommands/UpdateLabelAvailableFor/UpdateLabelAvailableForCommandHandlerTests.cs [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+    private UpdateLabelAvailableForCommandHandler _dut;
+    private UpdateLabelAvailableForCommand _command;
+
+    private Label _labelWithoutUsage;
+    private Label _labelForPunchPicture;
+    private readonly string _text = "A";
+    private readonly EntityTypeWithLabels _oldEntityTypeWithLabel = EntityTypeWithLabels.PunchPicture;
+    private readonly EntityTypeWithLabels _newEntityTypeWithLabel = EntityTypeWithLabels.PunchComment;
+
```
Comment (jsoi-eq) [inline src/tests/Equinor.ProCoSys.Completion.Command.Tests/LabelCommands/UpdateLabelAvailableFor/UpdateLabelAvailableForCommandHandlerTests.cs:29]: I have to admit I'm struggling to read and verify these tests, the cognitive load is a bit too much for me.

File: src/Equinor.ProCoSys.Completion.Infrastructure/Repositories/LabelRepository.cs [hunk 4]
```diff
@@ -14,10 +15,22 @@ public LabelRepository(CompletionContext context)
...
+        return DefaultQuery.Where(l => textsLowerCase.Contains(l.Text.ToLower()))
             .ToListAsync(cancellationToken);
     }
+
+    public async Task<Label> GetByTextAsync(string text, CancellationToken cancellationToken)
+    {
+        var textLowerCase = text.ToLower();
+        var label = await Set.Include(l => l.AvailableFor)
+            .Where(l => l.Text.ToLower() == textLowerCase).SingleOrDefaultAsync(cancellationToken);
+        if (label == null)
```
Comment (jsoi-eq) [inline src/Equinor.ProCoSys.Completion.Infrastructure/Repositories/LabelRepository.cs:30]: label is null

**Review by jsoi-eq — COMMENTED** [review]
Awaiting feedback

File: src/Equinor.ProCoSys.Completion.Command/LabelCommands/UpdateLabelAvailableFor/UpdateLabelAvailableForCommandHandler.cs (hunk 2 above)
Comment (janid1967) [reply src/Equinor.ProCoSys.Completion.Command/LabelCommands/UpdateLabelAvailableFor/UpdateLabelAvailableForCommandHandler.cs:36]: Better now @jsoi-eq ?

File: src/tests/Equinor.ProCoSys.Completion.Command.Tests/LabelCommands/UpdateLabelAvailableFor/UpdateLabelAvailableForCommandHandlerTests.cs (hunk 3 above)
Comment (janid1967) [reply src/tests/Equinor.ProCoSys.Completion.Command.Tests/LabelCommands/UpdateLabelAvailableFor/UpdateLabelAvailableForCommandHandlerTests.cs:29]: Better now @jsoi-eq ? Please resolve if OK

File: src/tests/Equinor.ProCoSys.Completion.Command.Tests/LabelCommands/UpdateLabelAvailableFor/UpdateLabelAvailableForCommandHandlerTests.cs [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+        _labelRepositoryMock.GetByTextAsync(_labelText, default).Returns(_labelAvailableForPunchPicture);
+        Assert.AreEqual(1, _labelAvailableForPunchPicture.AvailableFor.Count);
+        Assert.AreEqual(_punchPictureEntityType, _labelAvailableForPunchPicture.AvailableFor.ElementAt(0).EntityType);
+        var commandToMakeAvailableForNone = new UpdateLabelAvailableForCommand(_labelText, new List<EntityTypeWithLabel>());
+
+        // Act
+        await _dut.Handle(commandToMakeAvailableForNone, default);
+
+        // Assert
+        Assert.AreEqual(0, _labelNotAvailableToAnyEntityType.AvailableFor.Count);
```
Comment (jsoi-eq) [inline src/tests/Equinor.ProCoSys.Completion.Command.Tests/LabelCommands/UpdateLabelAvailableFor/UpdateLabelAvailableForCommandHandlerTests.cs:101]: Should this assert on _labelAvailableForPunchPicture instead?

Comment (janid1967) [reply src/tests/Equinor.ProCoSys.Completion.Command.Tests/LabelCommands/UpdateLabelAvailableFor/UpdateLabelAvailableForCommandHandlerTests.cs]: OOH, nice catch

**Review by jsoi-eq — APPROVED** [review]

**Pull request merged**